### ISSUE TITLE
ess per leapfrog step as the x axis in the cdf plot

### ISSUE THREAD

**Issue by mjhajharia** [mjhajharia/transforms#8]
I needed some quick clarification here: you get the ess after sampling is done, and the ess plot @sethaxen made is from concatenating ESS values from multiple sampling processes. Leapfrog steps are per iteration? so for the plot should I be looking at mean leapfrog step w.r.t every ESS value? I'm unsure

cc: @bob-carpenter

**Comment by bob-carpenter** [comment]
I agree with what @sethaxen said:  we get an ESS per leapfrog step value for each expectation (parameter x[n] and its square x[n]^2 at a care minimum) and can combine them only when they have the same posterior distribution (like in the symmetric Dirichlet case).  

One way to measure the overall efficacy is to measure the ESS of the log density lp__, as it's a nonlinear combination of the parameters.  Mixing lp__ is usually the limiting factor in HMC as it's evolution is tied to the momentum refresh.

**Comment by mjhajharia** [comment]
> > Leapfrog steps are per iteration? so for the plot should I be looking at mean leapfrog step w.r.t every ESS value?
> 
> Let's say we're sampling `n` parameters in `nruns` runs, each with `nchains` chains and `ndraws` draws per chain. For each run and parameter we get a single ESS value. Likewise, for each run, chain, and draw we get a number of leapfrog steps. I think what we would want to do is compute the ESS per total number of leapfrog steps computed for that given run (i.e sum number of leapfrog steps across all draws and chains for that run).
> 
> Then concatenating the ESS/total_leapfrog makes sense when the marginals should all follow the same distribution, which is the case whenever alpha is constant. For the case where alpha is a vector with multiple unique values, I think we should analyze each marginal separately.

Alright, I also think we should think more about how we would present non symmetric dirichlet, because when orders of magnitude for N increase it's not possible to put them in the paper or to look at it visually. I would say we take the mean ess for all x[i] and then check the variance of ess across them, if it's not significant we don't talk about it, if it is we try to explain why certain values of alpha created worse or better ess values

**Comment by mjhajharia** [comment]
ess/leapfrog plot for 100 repeated samples of 1000 draws

![image](https://user-images.githubusercontent.com/55153708/177868128-2e4c86ce-c6b9-42cd-814e-b07f0118a4b0.png)

will clean and push it in a bit

**Comment by bob-carpenter** [comment]
That plot looks great.  This looks like softmax is going to be the winner in that it's not too far behind stick breaking on average, but a lot better in stickbreaking's worst case.

How hard would it be to turn these into density plots rather than cdfs?  I find overlapping histograms hard to read, but density plots might be easier to read than these cdf plots.

**Comment by bob-carpenter** [comment]
The axes need labels, perhaps "cumulative probability" and "ESS / leapfrog".

**Comment by mjhajharia** [comment]
okay, that shouldn't be difficult, we can use arviz for consistency in plot design

**Comment by mjhajharia** [comment]
Arviz:

![image](https://user-images.githubusercontent.com/55153708/177882453-9e6e29c8-6f56-4684-8563-0b6fe6c18e00.png)

**Comment by bob-carpenter** [comment]
I don't know about everyone else, but I find this one easier to read than the CDFs.  For instance, it's easy to see there's substantial mass around zero for stick breaking in this one.

**Comment by mjhajharia** [comment]
> I don't know about everyone else, but I find this one easier to read than the CDFs.  For instance, it's easy to see there's substantial mass around zero for stick breaking in this one.

+1, not the most useful argument but these are visually better as well.

Also on an unrelated note: I was thinking of writing a small note under the ALR transform mentioning the CLR transform(standard softmax inverse) and say that this is not useful because..(not a bijection) and so we use ALR

**Comment by sethaxen** [comment]
> I don't know about everyone else, but I find this one easier to read than the CDFs. For instance, it's easy to see there's substantial mass around zero for stick breaking in this one.

For me the CDF one is also clear, but I'm not super opinionated that we use CDFs.

One thing to keep in mind is that there is not a unique softmax-augmented transform. The choice I made of $p_z(z)$ (see transforms section of the paper) is arbitrary. I don't have a principled argument to make for choosing a certain $p_z(z)$, but as I noted in a previous e-mail, in some informal tests the following seemed to work as well or better than softmax: 

https://github.com/mjhajharia/transforms/blob/e15856bb4f6c7a17bae04d728b0c1d7e230563a1/transforms/simplex/softmax-augmented/dirichlet_symmetric.stan#L14

with the choice $p=1/N$. I'd be curious to see how this compares in your benchmarks.

**Comment by bob-carpenter** [comment]
We can plot both.  I find the densities easier to follow in general, but the CDFs easier to read if I care about quantiles.  My mental autodiff isn't good enough to transform a CDF to a PDF in my head.

As I mentioned in email, I wrote a new section before I realized Seth had already written a derivation of the same thing.  I put it in terms of the standard additive log ratio (ALR) transform and it's much longer (1.5 pages vs. 0.5 pages), so hopefully easier to follow for those like me who didn't grow up with linear algebra and multivariate calc.  I'm afraid I can't find the p_z(z) bit.
